### PR TITLE
awareness of Sync

### PR DESCRIPTION
**Pull request by davirax**

### CODE REVIEW

File: sdk/src/main/scala/com/horizen/network/SidechainNodeViewSynchronizer.scala [hunk 1]
```diff
@@ -42,7 +76,155 @@ class SidechainNodeViewSynchronizer(networkControllerRef: ActorRef,
       }
   }
 
-  override protected def viewHolderEvents: Receive = onSyntacticallyFailedModifier orElse super.viewHolderEvents
+  private val onSemanticallyFailedModifier: Receive = {
+
+    //DIRAC TODO manage the StatusesMAp cancel the stuff down here
+
+    case SemanticallyFailedModification(mod, exception) =>
```
[the remark this answers is not in the document]
Comment (davirax) [reply sdk/src/main/scala/com/horizen/network/SidechainNodeViewSynchronizer.scala:74]: done

File: sdk/src/main/scala/com/horizen/network/SidechainNodeViewSynchronizer.scala [hunk 2]
```diff
@@ -42,7 +76,155 @@ class SidechainNodeViewSynchronizer(networkControllerRef: ActorRef,
...
+    setSyncAsDone orElse
+    betterNeighboursEvents orElse{
+      case a: Any => log.error("Strange input: " + a)
+    }
+  }
+
+  override def preStart(): Unit = {
+    super.preStart()
+    context.system.eventStream.subscribe(self, classOf[Events.NodeViewSynchronizerEvent])
+    context.system.eventStream.subscribe(self, classOf[SemanticallySuccessfulModifier[PMOD]])
```
[the remark this answers is not in the document]
Comment (davirax) [reply sdk/src/main/scala/com/horizen/network/SidechainNodeViewSynchronizer.scala]: done

File: sdk/src/main/scala/com/horizen/network/SidechainNodeViewSynchronizer.scala [hunk 3]
```diff
@@ -1,30 +1,64 @@
...
+import scala.reflect.ClassTag
 
-class SidechainNodeViewSynchronizer(networkControllerRef: ActorRef,
-                                    viewHolderRef: ActorRef,
-                                    syncInfoSpec: SidechainSyncInfoMessageSpec.type,
-                                    networkSettings: NetworkSettings,
-                                    timeProvider: NetworkTimeProvider,
-                                    modifierSerializers: Map[ModifierTypeId, ScorexSerializer[_ <: NodeViewModifier]])(implicit ec: ExecutionContext)
+
+class SidechainNodeViewSynchronizer[TX <: Transaction,
```
[the remark this answers is not in the document]
Comment (davirax) [reply sdk/src/main/scala/com/horizen/network/SidechainNodeViewSynchronizer.scala]: done

File: sdk/src/main/scala/com/horizen/network/SidechainDeliveryTracker.scala [hunk 4]
```diff
@@ -24,4 +40,63 @@ class SidechainDeliveryTracker(system: ActorSystem,
...
+// **********  these methods are Overridden for keeping info before they disappear
+
+  override def setReceived(id: ModifierId, sender: ConnectedPeer): Unit = {
+    log.info(s"moving to my requestedTo before to setReceived $id from peer: $sender")
+    beenRequestedTo.put(id,sender)
+    // work as before
+    super.setReceived(id,sender)
+  }
+
+  override def setHeld(id: ModifierId): Unit ={
```
[the remark this answers is not in the document]
Comment (davirax) [reply sdk/src/main/scala/com/horizen/network/SidechainDeliveryTracker.scala:52]: done by differentiating ths calls, now there i a method setBlockHeld, that makes his checks and after calls super.setHeld

File: sdk/src/main/scala/com/horizen/network/SidechainNodeViewSynchronizer.scala [hunk 5]
```diff
@@ -42,7 +76,155 @@ class SidechainNodeViewSynchronizer(networkControllerRef: ActorRef,
...
+  protected def betterNeighboursEvents: Receive = {
+    case BetterNeighbourAppeared =>
+      log.info("received a good neighbour ")
+
+    case NoBetterNeighbour =>
+      log.info("No better neighbours around ")
+  }
+
+  def wasRequested(pmod: PersistentNodeViewModifier): Option[ConnectedPeer] = {
+    val peerToWhichIRequested= deliveryTracker.modHadBeenRequestedFromPeer(pmod.id,pmod.modifierTypeId)
```
[the remark this answers is not in the document]
Comment (davirax) [reply sdk/src/main/scala/com/horizen/network/SidechainNodeViewSynchronizer.scala]: done

File: sdk/src/main/scala/com/horizen/network/SidechainSyncTracker.scala [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+                            networkSettings: NetworkSettings,
+                            timeProvider: TimeProvider)(implicit ec: ExecutionContext)
+  extends SyncTracker(nvsRef: ActorRef,
+    context: ActorContext,
+    networkSettings: NetworkSettings,
+    timeProvider: TimeProvider) {
+
+  private var olderStatusesMap = mutable.Map[ConnectedPeer, SidechainSyncStatus]()
+  var betterNeighbourHeight = 0
+  var myHeight = 0 // DIRAC TODO update first on processSync after every succesfulModifier applied
```
[the remark this answers is not in the document]
Comment (davirax) [reply sdk/src/main/scala/com/horizen/network/SidechainSyncTracker.scala]: agreed to leave it

File: sdk/src/main/scala/com/horizen/network/SidechainSyncTracker.scala [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+  def updateSyncStatus(peer: ConnectedPeer, syncStatus: SidechainSyncStatus): Unit = {
+    super.updateStatus(peer,syncStatus.historyCompare)
+    if (syncStatus.historyCompare == Older){
+      log.info(s"updating syncStatus , height = ${syncStatus.otherNodeDeclaredHeight}")
+      olderStatusesMap += peer -> syncStatus
+      // DIRAC not sure if correct, just to have the best height we got through the sync phase
```
[the remark this answers is not in the document]
Comment (davirax) [reply sdk/src/main/scala/com/horizen/network/SidechainSyncTracker.scala]: agreeed

File: sdk/src/main/scala/com/horizen/network/SidechainNodeViewSynchronizer.scala [hunk 8]
```diff
@@ -42,7 +76,155 @@ class SidechainNodeViewSynchronizer(networkControllerRef: ActorRef,
...
+          // Ban both mod.id and peer
+          deliveryTracker.setInvalid(mod.id).foreach(penalizeMisbehavingPeer)
+      }
+  }
+
+
+
+  private val onSemanticallySuccessfulModifier: Receive = {
+    case SemanticallySuccessfulModifier(pmod) =>
+      val aPeer = wasRequested(pmod)
```
[the remark this answers is not in the document]
Comment (davirax) [reply sdk/src/main/scala/com/horizen/network/SidechainNodeViewSynchronizer.scala]: done

File: sdk/src/main/scala/com/horizen/network/SidechainDeliveryTracker.scala [hunk 9]
```diff
@@ -24,4 +40,63 @@ class SidechainDeliveryTracker(system: ActorSystem,
         None
     }
   }
+
+// **********  these methods are to be called from SC_NVS for checking the origin of the applied mod
+
+  def modHadBeenRequestedFromPeer( id:ModifierId,typeId: ModifierTypeId ):Option[ConnectedPeer] = {
+    log.info(s"modHadBeenRequestedFromPeer: I had Received <$id> from the peer: ${beenReceivedFrom.get(id)}")
+    beenRequestedTo.get(id)
```
[the remark this answers is not in the document]
Comment (davirax) [reply sdk/src/main/scala/com/horizen/network/SidechainDeliveryTracker.scala]: done

File: sdk/src/main/scala/com/horizen/network/SidechainDeliveryTracker.scala [hunk 10]
```diff
@@ -24,4 +40,63 @@ class SidechainDeliveryTracker(system: ActorSystem,
...
+
+  override def setReceived(id: ModifierId, sender: ConnectedPeer): Unit = {
+    log.info(s"moving to my requestedTo before to setReceived $id from peer: $sender")
+    beenRequestedTo.put(id,sender)
+    // work as before
+    super.setReceived(id,sender)
+  }
+
+  override def setHeld(id: ModifierId): Unit ={
+    val peerFromWhichIReceived = received.get(id).get
```
[the remark this answers is not in the document]
Comment (davirax) [reply sdk/src/main/scala/com/horizen/network/SidechainDeliveryTracker.scala]: done

**Comment by i-Alex** [comment]
Not actual

**Pull request closed without merging**